### PR TITLE
INFO-3 Subscription channels remain open after token expiration

### PR DESCRIPTION
**Pull request by lukasmittag**
This fixes that a subscription channel stays open after the token of the subscriber has expired. We need to decide if we want to fix this. Currently it is only fixed for ```ChangeSubscription``` which the new API uses. If we want to do it for the old query subscription as well it can  be done similiar. To test it take a jwt token of kuksa-val/jwt/ and use a timestamp that expires in the near future -> go to https://www.epochconverter.com/ for timestamps ;)

It is not that pretty in code I think so do not now if we want to fix this in this way/at all. One observation I made was that even if the channel is kept open somehow if you update the token new updates do not come in in the databroker-cli again. Haven't figured out why yet.

### CODE REVIEW

**Comment by lukasmittag** [comment]
and I opted for closing if a new notification would be sent out so no general repeated check. This means if you do not update soem things channel can stay open a while.

**Comment by rafaeling** [comment]
Code looks fine to me, tested with python sdk and the channel closes correclty

**Review by erikbosch — APPROVED** [review]
Looks ok for me. But I did not really understood one of the comment, see below. Is that one still valid, or do you think the code is pretty enough (kenough) now?

"It is not that pretty in code I think so do not now if we want to fix this in this way/at all."

**Comment by lukasmittag** [comment]
I just thought that it's maybe not the ideal way code wise. But since @argerus had nothing to complain yet :) From my side we can merge.

File: kuksa_databroker/databroker/src/broker.rs [hunk 1]
```diff
@@ -626,10 +627,14 @@ impl Subscriptions {
...
 
-        for sub in &self.change_subscriptions {
+        for sub in &mut self.change_subscriptions {
             match sub.notify(changed, db).await {
                 Ok(_) => {}
-                Err(err) => error = Some(err),
+                Err(err) => {
+                    error = Some(err);
+                    let taken_sender = take(&mut sub.sender);
+                    drop(taken_sender);
```
Comment (argerus) [inline kuksa_databroker/databroker/src/broker.rs:636]: I don't understand the logic of what is going on here at all... No need to use `mem::take` (to replace the sender with a "default" value?) only to immediately `drop` it.

Just set it to `None`, which I assume was your intention when changed it to `Option<Sender>`, right?

```suggestion
                    sub.sender = None:
```

**Review by argerus — CHANGES_REQUESTED** [review]
It would probably be a lot easier to treat expired permissions the same way we treat channels that have been closed:

e.g. add a check to `Subscriptions::cleanup()` that will remove all subscriptions with expired permissions. This will be triggered both periodically, and whenever a `SubscriptionError` is returned when when calling notify.

File: kuksa_databroker/databroker/src/broker.rs (hunk 1 above)
Comment (argerus) [reply kuksa_databroker/databroker/src/broker.rs:636]: (or revert this part and drop subscriptions the same way it's done when a closed channel is detected, see general comment)

Comment (lukasmittag) [reply kuksa_databroker/databroker/src/broker.rs:636]: your second suggestion makes the most sense from my POV. I will handle it in cleanup method if it gets called periodically and on error. Thanks for the remarks :)

Comment (argerus) [reply kuksa_databroker/databroker/src/broker.rs:636]: Awesome :wink:

**Comment by erikbosch** [comment]
Databroker has been migrated to https://github.com/eclipse-kuksa/kuksa-databroker.
Please open a new pull request in that repo.

**Pull request closed without merging**